### PR TITLE
Add Flex/Bison parser

### PR DESCRIPTION
**Pull request by makaimann**
This PR adds an `SmtLibReader` class with virtual methods for common SMT-LIB commands. This way, a user can inherit the class and perform their own actions on standard commands. By default, it just executes the commands on the provided solver.

Currently supports:
* Bool/Int/Real/BV/Array/UF
* quantifiers
* let bindings
* declare-fun/declare-const/define-fun
* set-logic/set-option/set-info
* check-sat/check-sat-assuming/push/pop/exit

The infrastructure is based on the [calc++ example](https://www.gnu.org/software/bison/manual/html_node/A-Complete-C_002b_002b-Example.html).

I'm new to Flex/Bison so please let me know if there are any inefficiencies/broken conventions that you notice. Of course, trying to balance making it nice vs. getting it done soon so it's not going to be perfect. One particularly ugly part right now is here: 
https://github.com/makaimann/smt-switch/blob/9b7fe79103648da4d5e7fc4ea0bf973abfd0e4a7/src/smtlibscanner.l#L72

Basically, there can be multiline string literals in set-info, or in a pipe-quoted symbol, e.g.:
```
(set-logic QF_LIA)
(declare-const |a
                multi-line
                symbol|
                Int)
```

Currently, the lexer combines that into one token. However, this can mess up error reporting since the line number is off. So that linked code just increments the line number accordingly. I did briefly try splitting them into one token per line and recombining them in the bison parser but ran into some issues. I could probably get it working if you think that's better. Let me know if you think that's a lot better than the current solution.

### CODE REVIEW

**Comment by makaimann** [comment]
initial-smt-switch-parser-results

Here are the initial parser results. Note: incremental benchmarks have more solved than the total since each sat/unsat counts as a solve. I have no idea what's going on with the QF_AUFLIA benchmarks. It could be a parser issue, but since all the others seem to be fine, I'm guessing it's actually a difference in the default options. I'll look into that more today.

**Comment by makaimann** [comment]
Btw, the tests above were ran with an older version. I'll run tests again soon with the latest.  Looking into cvc4's default options first.

**Comment by makaimann** [comment]
All right, I found the slowdown. The `crash3.smt2` benchmark in the `safari` set was very large and made frequent use of `define-funs`. My `define-fun` handling uses naive substitution, and the substitution operation was slow for CVC4. Fixed with: https://github.com/makaimann/smt-switch/pull/184. Here are the latest results where smt-switch wins in most categories:

parser-updated

**Comment by barrettcw** [comment]
Awesome!

**Review by ahmed-irfan — COMMENTED** [review]
Looks great! 

I guess the get commands (`get-unsat-core, get-value, get-model` etc) are left for future work. RIght?

**Comment by ahmed-irfan** [comment]
> All right, I found the slowdown. The `crash3.smt2` benchmark in the `safari` set was very large and made frequent use of `define-funs`. My `define-fun` handling uses naive substitution, and the substitution operation was slow for CVC4. Fixed with: #184. Here are the latest results where smt-switch wins in most categories:
> 
> parser-updated

That is indeed great!

Interestingly, smt-switch-cvc4 solves way more in the QF_AUFLIA

**Comment by makaimann** [comment]
Thanks for the review!
> I guess the get commands (`get-unsat-core, get-value, get-model` etc) are left for future work. Right?

Yeah that's right -- I haven't thought about those much yet. `get-value` would be easy if we need that. I think `get-unsat-core` and `get-model` would take a bit more thought.

**Comment by makaimann** [comment]
> Interestingly, smt-switch-cvc4 solves way more in the QF_AUFLIA

Right, so that's actually entirely from the single `crash3.smt2` benchmark. It has a huge number of `check-sat` calls (1,109,912 total) and they're all very quick (at least all the ones that we get to in the time limit). So I think there's a noticeable difference there just because flex/bison is faster than ANTLR and there's very little time in the solver.

File: src/smtlib_reader.cpp [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+{
+  Result r = solver_->check_sat();
+  cout << r << endl;
+  return r;
+}
+
+Result SmtLibReader::check_sat_assuming(const TermVec & assumptions)
+{
+  Result r = solver_->check_sat_assuming(assumptions);
+  cout << r << endl;
```
[the remark this answers is not in the document]
Comment (makaimann) [reply src/smtlib_reader.cpp:82]: Good question. So for the binary we need it somewhere to print the result. But does it belong in this class or not... I'm not sure. What do you think? I don't want to put it in the actual parser. I chose to put it here so that if you don't want it to print, you can just override the function. But another option is to _not_ print in this class and add it in a derived class.

File: src/smtlib_reader.cpp [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+  if (it == defs_.end())
+  {
+    throw SmtException("Unknown function: " + defname);
+  }
+
+  Term def = it->second;
+
+  if (!num_args)
+  {
+    def;
```
[the remark this answers is not in the document]
Comment (makaimann) [reply src/smtlib_reader.cpp]: Lol what is happening there? Thanks for catching that! Meant to be a return, fixed with e7e7531eb20ecb3d74762b964f7b34192a188aff. Not an actual bug because all the code afterwards still works, but if there are no arguments then we don't need to substitute at all.

[the remark this answers is not in the document]
Comment (makaimann) [reply src/smtlib_reader.cpp]: And actually, on second thought that code should never run at all. Because `apply_define_fun` should only get run with arguments. I'll change that into an assertion.

File: src/smtlibscanner.l [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+                          }
+                        }
+                        loc.step();
+                        return yy::parser::make_QUOTESTRING(noquotes, loc);
+                      }
+
+[0-9]+                { return yy::parser::make_NAT(yytext, loc); }
+[0-9]+\.[0-9]+        { return yy::parser::make_FLOAT(yytext, loc); }
+#b[01]+               { yytext=yytext+2; return yy::parser::make_BITSTR(yytext, loc); }
+#x[0-1A-F]+           { yytext=yytext+2; return yy::parser::make_HEXSTR(yytext, loc); }
```
[the remark this answers is not in the document]
Comment (makaimann) [reply src/smtlibscanner.l]: Oops, definitely. Good catch! Fixed with ee87e56a30f561c9fd09670b8cbe77311a1ca5d4.

File: src/smtlibparser.yy [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+  LP SETLOGIC SYMBOL RP
+  {
+    drv.set_logic($3);
+  }
+  | LP SETOPT KEYWORD SYMBOL RP
+  {
+    drv.set_opt($3, $4);
+  }
+  | LP SETINFO KEYWORD number_or_string RP
+    {
```
[the remark this answers is not in the document]
Comment (makaimann) [reply src/smtlibparser.yy:130]: Fixed with 6e2020d6d845390f9b5387ce6c94927b49e69140.

File: src/smtlibparser.yy [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+s_expr:
+  atom
+  {
+    $$ = $1;
+  }
+  | LP operator s_expr_list RP
+  {
+    // special-case for MINUS
+    // needs to be negate if only one argument
+    // TODO: might be a more elegant way to handle this
```
[the remark this answers is not in the document]
Comment (makaimann) [reply src/smtlibparser.yy:208]: That's a good idea, but I ran into a shift-reduce conflict. How would you disambiguate between `(- x)` and `(- x y)` in the grammar? The issue with the way I set it up is that I have `LP operator s_expr_list RP`, and `s_expr_list` can be a single `s_expr`, and `operator` needs to include `MINUS`.

File: src/smtlibparser.yy [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+   }
+;
+
+
+%%
+
+void yy::parser::error (const location_type& l, const std::string& m)
+{
+  cerr << l << ": " << m << endl;
+}
```
[the remark this answers is not in the document]
Comment (makaimann) [reply src/smtlibparser.yy]: Addressed with 5113195dd697530fc15329fa1eec236ede5b4a44.

File: src/smtlibparser.yy (hunk 5 above)
[the remark this answers is not in the document]
Comment (ahmed-irfan) [reply src/smtlibparser.yy:208]: ok, let's not do it then.

File: src/smtlib_reader.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (amaleewilson) [reply src/smtlib_reader.cpp:82]: I think it's fine either way.

**Review by amaleewilson — APPROVED** [review]
Looks good to me

**Review by ahmed-irfan — APPROVED** [review]
LGTM.

Suggestion: please open issues for the todos, like supporting getter functions in the parser.

**Comment by makaimann** [comment]
Thanks for the reviews! Made an issue for the getters: https://github.com/makaimann/smt-switch/issues/190.
